MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,417 @@
+    config: config object
+    rng: jax.prng key
+    mesh: jax.devices() mesh
+    checkpoint_manager: an Orbax checkpointing.CheckpointManager object
+
+  Returns:
+    state: the initialized train state
+    state_mesh_annotations: the mesh annotations for the train state
+  """
+
+  unboxed_abstract_state, state_mesh_annotations = get_abstract_state(model, tx, config, rng, mesh)
+
+  # Initialization
+  with nn_partitioning.axis_rules(config.logical_axis_rules):
+    state, raw_params = checkpointing.load_state_if_possible(checkpoint_manager,
+                                                config.load_parameters_path,
+                                                config.load_from_other_directory,
+                                                config.load_from_other_directory_step,
+                                                unboxed_abstract_state,
+                                                mesh,
+                                                state_mesh_annotations)
+
+    state_mesh_shardings = jax.tree_map(
+        lambda p: jax.sharding.NamedSharding(mesh, p), state_mesh_annotations)
+    if not state:
+      init_train_state_partial = functools.partial(init_train_state, model, tx, config)
+      state = jax.jit(
+          init_train_state_partial,
+          in_shardings=None,
+          out_shardings=state_mesh_shardings
+      )(rng)
+      if raw_params: # If we loaded a partial state, we need to merge it.
+        state = state.replace(params = raw_params)
+    raw_params = None
+
+  state = unbox_logicallypartioned_trainstate(state)
+  return state, state_mesh_annotations
+
+def create_device_mesh(config, devices=jax.devices(), logging=True):
+  """Creates a device mesh with each slice in its own data parallel group. If there is only one slice, uses two replicas """
+  num_devices = len(devices)
+  try:
+    num_slices = 1 + max([d.slice_index for d in devices])
+  except:
+    num_slices = 1
+  num_devices_per_slice = num_devices//num_slices
+  max_logging.log(f"Devices: {devices} (num_devices: {num_devices})")
+  assert len(devices) > 1, "You must have at least two devices"
+
+  multi_slice_env = num_slices > 1
+
+  dcn_parallelism = [config.dcn_data_parallelism, config.dcn_fsdp_parallelism, config.dcn_tensor_parallelism]
+  ici_parallelism = [config.ici_data_parallelism, config.ici_fsdp_parallelism, config.ici_tensor_parallelism]
+
+  # Find possible unspecified parallelisms
+  dcn_parallelism = fill_unspecified_mesh_axes(dcn_parallelism, num_slices, 'DCN')
+  ici_parallelism = fill_unspecified_mesh_axes(ici_parallelism, num_devices_per_slice, 'ICI')
+
+  if multi_slice_env:
+    mesh = mesh_utils.create_hybrid_device_mesh(ici_parallelism, dcn_parallelism, devices)
+  else:
+    mesh = mesh_utils.create_device_mesh(ici_parallelism, devices)
+
+  if logging:
+    max_logging.log(f"Decided on mesh: {mesh}")
+
+  return mesh
+
+def fill_unspecified_mesh_axes(parallelism_vals, target_product, parallelism_type):
```

Comment:
Ditto above as create_device_mesh, its possible this guy wants to be shared, in which case in can be moved when someone else wants it